---
title: "HIT-GLIO sample1 myeloid clusters annotation"
author: "MF"
date: "`r Sys.Date()`"
# output: powerpoint_presentation
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
      number_sections: true
---


```{r, echo=FALSE, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(Seurat)
  library(ggplot2)
  library(Matrix)
  library(tidyverse)
  library(clustree)
  library(reticulate) # load packages required by leiden algo
  reticulate::use_condaenv("/mnt/sda4/conda/envs/R")
  library(viridis) # colors
  options(future.globals.maxSize = 3.0 * 1e9) # 3GB
  library(dittoSeq)
  library(AUCell)
  library(GSVA)
  library(msigdbr)
  library(Nebulosa)
  library(monocle3)
  library(SeuratExtend)
  # library(escape)
  library(fgsea)
  library(DT)
})
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# seu_singlet_myelo <- readRDS("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/07-Feb-2025_seu_singlet_myelo.rds") # nolint

seu_singlet_myelo <- readRDS("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/10-Feb-2025_seu_singlet_myelo_silhouette.rds") # nolint

group_by <- "wsnn_res.0.4"

source("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/scripts/functions.R")
source("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/scripts/00_Global_vars.R")
source("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/scripts/cell_features.R")

set.seed(123)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
nclusters <- length(unique(seu_singlet_myelo@meta.data$wsnn_res.0.4))
myelo_cluster_annotation <- rep("unannotated", nclusters)
names(myelo_cluster_annotation) <- levels(
  seu_singlet_myelo@meta.data$wsnn_res.0.4
)

myelo_cluster_annotation["1"] <- "Mg-TAM (M1) (CD14+, CD74+) [C1]"
myelo_cluster_annotation["2"] <- "Mg-TAM (M2) (CD14+, CD74-) [C2]"
myelo_cluster_annotation["3"] <- "Microglia [C3]"
myelo_cluster_annotation["4"] <- "Microglia [C3]"
myelo_cluster_annotation["5"] <- "Mo-TAM (1) [C4]"
myelo_cluster_annotation["6"] <- "Monocytes/MDSC (CD11b+, CD14-) [C5]"
myelo_cluster_annotation["7"] <- "Neutrophils [C6]"
myelo_cluster_annotation["8"] <- "Mo-TAM (2) [C7]"
myelo_cluster_annotation["9"] <- "Doublets [C8]"
myelo_cluster_annotation["10"] <- "cDC1 [C9]"
myelo_cluster_annotation["11"] <- "B-Cell [C10]"
myelo_cluster_annotation["12"] <- "pDC [C11]"

seu_singlet_myelo <- SetIdent(
  seu_singlet_myelo,
  value = seu_singlet_myelo@meta.data$wsnn_res.0.4
)
seu_singlet_myelo <- RenameIdents(seu_singlet_myelo, myelo_cluster_annotation)
seu_singlet_myelo@meta.data$manual_anno <- Idents(seu_singlet_myelo)
```

# Final Dimplot of myeloid clusters
```{r, echo=FALSE, message=FALSE, warning=FALSE}
SeuratExtend::DimPlot2(
  seu_singlet_myelo,
  reduction = "wnn.umap",
  label = TRUE,
  label.color = "black",
  index.title = "C",
  repel = TRUE,
  pt.size = 1.3,
  box = TRUE,
  theme = NoAxes()
) + theme_umap_arrows()

DimPlot2(
  seu_singlet_myelo,
  group.by = "MULTI_ID",
  reduction = "wnn.umap",
  label = TRUE,
  pt.size = 1.3,
  theme = NoAxes()
) + theme_umap_arrows()
```

# CellTypist to predict labels
https://www.celltypist.org/
prediction is based on Immune_ALL_high reference
```{r, echo=FALSE, message=FALSE, warning=FALSE}
# annot <- read.csv("/mnt/sda4/cell_typist_predicted_labels.csv")
# colnames(annot)
# rownames(annot) <- annot$cell

# seu_singlet_myelo <- AddMetaData(
#   seu_singlet_myelo,
#   metadata = annot
# )

# DimPlot(
#   seu_singlet_myelo,
#   reduction = "umap.myelo.pca.sct",
#   group.by = "predicted_labels",
#   label = TRUE,
#   repel = TRUE
# )
```

# Diff.Expr. Cluster X vs REST
Make lognorm for RNA assay. Subset 100 cells.
```{r, echo=FALSE, message=FALSE, warning=FALSE}
DefaultAssay(seu_singlet_myelo) <- "RNA"

# seu_singlet_myelo <- seu_normalize_var_scale_pca(
#   seu_singlet_myelo,
#   normalize = "log",
#   cluster.name = "cluster.log.myelo",
#   run_pca = TRUE,
#   pca.reduction.name = "pca.myelo.log",
#   umap.reduction.name = "umap.myelo.pca.log",
#   dims = 1:30,
#   k.param = 30,
#   algorithm = 4,
#   resolution = 1.0,
#   group.singletons = FALSE
# )
# keep sct idents on; log is only for DE calculations
# Idents(seu_singlet_myelo) <- "cluster.sct.myelo"
# Idents(seu_singlet_myelo) <- "wsnn_res.0.4"

# DimPlot(seu_singlet_myelo, reduction = "umap.myelo.pca.sct", label = TRUE)
# DimPlot(seu_singlet_myelo, reduction = "wnn.umap", label = TRUE)

# for the reference, this is how log clustering looks like
# DimPlot(seu_singlet_myelo, reduction = "umap.myelo.pca.log", label = TRUE)
```

## subset even number of cells
```{r, echo=FALSE, message=FALSE, warning=FALSE}
table(seu_singlet_myelo@active.ident)

cluster_ids <- Idents(seu_singlet_myelo)

# Initialize an empty list to store the subsetted cells
subset_cells <- list()

# Loop through each cluster and sample cells
for (cluster in unique(cluster_ids)) {
  cells_in_cluster <- WhichCells(seu_singlet_myelo, idents = cluster)
  sampled_cells <- sample(cells_in_cluster, size = 40)
  # size = min(100, length(cells_in_cluster)) # if no of cell smaller than 100, select max cells in cluster # nolint
  subset_cells[[cluster]] <- sampled_cells
}

# Combine the sampled cells into a single vector
subset_cells_all <- unlist(subset_cells)

# Subset the Seurat object
seurat_subset <- subset(seu_singlet_myelo, cells = subset_cells_all)

seurat_subset <- NormalizeData(seurat_subset)
seurat_subset <- ScaleData(seurat_subset)

table(seurat_subset@active.ident)
```

## Compute differentiall expression
```{r, echo=FALSE, message=FALSE, warning=FALSE}
markers_genes <- FindAllMarkers(
  seurat_subset,
  log2FC.threshold = 0.2,
  test.use = "MAST",
  min.pct = 0.1,
  min.diff.pct = 0.2,
  only.pos = TRUE,
  # max.cells.per.ident = 50, this subsamples randomly every time, better to subsample on your own. # nolint
  assay = "RNA"
)
```

## draw barlopts with top25
```{r, echo=FALSE}
top25 <- markers_genes %>%
  group_by(cluster) %>%
  top_n(-25, p_val_adj)
```

```{r}
for (i in unique(top25$cluster)) {
  # print genes protted on barplots (top25 genes in each cluster)
  omg <- top25 |>
    filter(cluster == i)
  print(omg$gene)

  barplot(
    sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], FALSE),
    horiz = TRUE,
    las = 1,
    main = paste0(i, " vs. rest"),
    border = "white",
    yaxs = "i"
  )
  abline(v = c(0, 0.25), lty = c(1, 2))
}
```

## make heatmap with top 10 genes
```{r, message=FALSE, echo=FALSE, warning=FALSE, error=FALSE, fig.width=10, fig.height=17}
top_x <- markers_genes %>%
  group_by(cluster) %>%
  slice_min(p_val_adj, n = 10, with_ties = FALSE)

seurat_subset_heatmap <- seurat_subset

seurat_subset_heatmap <- ScaleData(
  seurat_subset_heatmap,
  features = as.character(unique(top_x$gene)), assay = "RNA"
)

DoHeatmap(
  seurat_subset_heatmap,
  features = as.character(unique(top_x$gene)),
  # group.by = "cluster.sct.myelo", assay = "RNA"
  group.by = "ident", assay = "RNA"
)
```

## make dotplot with top 10 (same viz. as heatmap)
```{r, echo=FALSE, message=FALSE, fig.width=10, fig.height=20}
DotPlot(
  seurat_subset_heatmap,
  features = rev(as.character(unique(top_x$gene))), group.by = "manual_anno", assay = "RNA"
) + coord_flip() + mycolor2 + theme(
  axis.text.x = element_text(angle = 90, hjust = 1)
)
```

## validate top10 gene expression on vlnplots
```{r, echo=FALSE, message=FALSE, fig.width=12, fig.height=15}
for (i in unique(top_x$cluster)) {
  top_x_cluster <- top_x |>
    filter(cluster == i)

  print(paste0("cluster: ", i))

  print(VlnPlot(
    seu_singlet_myelo,
    features = top_x_cluster$gene
  ) + patchwork::plot_annotation(
    paste0("Cluster ", i, " top genes"),
    theme = theme(plot.title = element_text(size = 25))
  ))
}
```

## show topgenes of each cluster in feature and density plots
### FeaturePlots
```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=20, fig.height=15}
for (i in unique(top_x$cluster)) {
  top_x_cluster <- top_x |>
    filter(cluster == i)

  feat1 <- FeaturePlot(
    seu_singlet_myelo,
    as.vector(top_x_cluster$gene),
    pt.size = 1,
    label = FALSE,
    reduction = "wnn.umap"
  ) + patchwork::plot_annotation(
    title = paste0("Top genes in Cluster ", i),
    theme = theme(plot.title = element_text(size = 25))
  ) & mycolor2 & NoAxes()
  print(feat1)

  dens1 <- plot_density(
    seu_singlet_myelo,
    as.vector(top_x_cluster$gene),
    size = 1.5,
    joint = TRUE,
    reduction = "wnn.umap"
  ) + patchwork::plot_annotation(
    title = paste0("Top genes in Cluster ", i),
    theme = theme(plot.title = element_text(size = 25))
  ) & NoAxes()
  print(dens1)
}
```

# Generate scores of marker gene sets using AUCell
## Run AUCEll
```{r, echo = FALSE, warning=FALSE}
expr_matrix <- as.matrix(
  GetAssayData(seu_singlet_myelo, layer = "data", assay = "RNA")
)

gene_rankings <- AUCell_buildRankings(expr_matrix, plotStats = FALSE)
```

### wang2024_hypoxic_macrophages
#### FeaturePlot of AUCell scores
```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=20, fig.height=25}
print(wang2024_hypoxic_macrophages)

auc_scores <- AUCell_calcAUC(
  wang2024_hypoxic_macrophages,
  gene_rankings
)

auc_matrix <- as.data.frame(t(getAUC(auc_scores)))
seu_singlet_myelo <- AddMetaData(seu_singlet_myelo, metadata = auc_matrix)

images <- map(
  names(wang2024_hypoxic_macrophages),
  ~ FeaturePlot(
    seu_singlet_myelo,
    features = .x,
    pt.size = 1,
    label = FALSE,
    reduction = "wnn.umap"
  ) & mycolor2 & NoAxes()
)

stacked_images <- patchwork::wrap_plots(images, ncol = 3)
plot1 <- stacked_images + patchwork::plot_annotation(
  title = "Wang2024 hypoxic macrophages",
  theme = theme(plot.title = element_text(size = 25))
)
print(plot1)
```

#### Density Plot of genes
```{r, echo = FALSE, fig.width=20, fig.height=20}
wang2024_hypoxic_macrophages_cleaned <- map(
  wang2024_hypoxic_macrophages,
  function(x) x[x %in% rownames(seu_singlet_myelo)]
)

imap( # indexed map
  wang2024_hypoxic_macrophages_cleaned,
  ~ plot_density(
    seu_singlet_myelo,
    features = .x,
    size = 1.5,
    joint = TRUE,
    reduction = "wnn.umap"
  ) + patchwork::plot_annotation(
    title = .y, # name of named vector in a list
    theme = theme(plot.title = element_text(size = 25))
  ) & NoAxes()
)
```

### Greenwald2024_Spatial transcriptomics
#### FeaturePlot of AUCell scores
```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=12}
print(greenwald2024)

auc_scores <- AUCell_calcAUC(
  greenwald2024,
  gene_rankings
)

auc_matrix <- as.data.frame(t(getAUC(auc_scores)))
seu_singlet_myelo <- AddMetaData(seu_singlet_myelo, metadata = auc_matrix)

images <- map(
  names(greenwald2024),
  ~ FeaturePlot(
    seu_singlet_myelo,
    features = .x,
    pt.size = 1,
    label = FALSE,
    reduction = "wnn.umap"
  ) & mycolor2 & NoAxes()
)

stacked_images <- patchwork::wrap_plots(images, ncol = 3)
plot1 <- stacked_images + patchwork::plot_annotation(
  title = "Greenwald2024 Integrative spatial analysis reveals a multi-layered organization of glioblastoma",
  theme = theme(plot.title = element_text(size = 25))
)
print(plot1)
```

### Beck2021 pHGG metapropgrams
```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=13, fig.height=14}
print(liu2022)

auc_scores <- AUCell_calcAUC(
  liu2022,
  gene_rankings
)

auc_matrix <- as.data.frame(t(getAUC(auc_scores)))
seu_singlet_myelo <- AddMetaData(seu_singlet_myelo, metadata = auc_matrix)

images <- map(
  names(liu2022),
  ~ FeaturePlot(
    seu_singlet_myelo,
    features = .x,
    pt.size = 1,
    label = FALSE,
    reduction = "wnn.umap"
  ) & mycolor2 & NoAxes()
)

stacked_images <- patchwork::wrap_plots(images, ncol = 3)
plot1 <- stacked_images + patchwork::plot_annotation(
  title = "AUCell scores of Liu2022 The Landscape of Tumor Cell States and Spatial Organization in H3-K27M Mutant Diffuse Midline Glioma",
  theme = theme(plot.title = element_text(size = 25))
)
print(plot1)
```



### VlnPlot of AUC scores of microglia/macrophages core markers
```{r, fig.width=10, fig.height=7}
# Idents(seu_singlet_myelo) <- "merged_to_one"

mgmc <- list(
  microglia1 = c("HEXB", "TMEM119", "P2RY12"),
  microglia2 = c("CX3CR1", "P2RY12", "P2RY13", "SELPLG"),
  macrophages1 = c("CD163", "CD74", "TGFBI", "IFITM2", "IFITM3", "F13A1", "NPC2", "FTH1") # nolint
)
sapply(mgmc, function(x) x) # print

auc_scores <- AUCell_calcAUC(
  mgmc,
  gene_rankings
)
gene_rankings <- AUCell_buildRankings(expr_matrix, plotStats = FALSE)

auc_matrix <- as.data.frame(t(getAUC(auc_scores)))
seu_singlet_myelo <- AddMetaData(seu_singlet_myelo, metadata = auc_matrix)
```

```{r, fig.width=10, fig.height=15}
VlnPlot(
  seu_singlet_myelo,
  features = c(
    "microglia1",
    "microglia2",
    "macrophages1"
  )
)
```

### VlnPlot of expression of individual genes of microglia/macropahges core markers
```{r, fig.width=10, fig.height=12}
imap(
  mgmc,
  ~ VlnPlot(
    seu_singlet_myelo,
    features = .x
  ) + patchwork::plot_annotation(
    title = .y,
    theme = theme(plot.title = element_text(size = 25))
  )
)
```


### FeaturePlot of AUC scores of microglia/macrophages core markers
```{r, echo=FALSE, message=FALSE, fig.width=12, fig.height=6}
FeaturePlot(
  seu_singlet_myelo,
  features = c(
    "microglia1",
    "microglia2",
    "macrophages1"
  ),
  pt.size = 1,
  label = FALSE,
  reduction = "wnn.umap"
) & mycolor2 & NoAxes()
```

### DensityPlot of microglia/macrophages genes
```{r, echo=FALSE, message=FALSE, fig.width=20, fig.height=20}
imap( # indexed map
  mgmc,
  ~ plot_density(
    seu_singlet_myelo,
    features = .x,
    size = 1.5,
    joint = TRUE,
    reduction = "wnn.umap"
  ) + patchwork::plot_annotation(
    title = .y, # name of named vector in a list
    theme = theme(plot.title = element_text(size = 25))
  ) & NoAxes()
)
```

### Heatmap z-score of microglia/macrophages core markers
```{r, echo=FALSE, message=FALSE, fig.width=8, fig.height=7}
features <- c(mgmc$microglia1, mgmc$microglia2, mgmc$macrophages1)

seu_singlet_myelo_heatmap <- seu_singlet_myelo

seu_singlet_myelo_heatmap <- ScaleData(
  seu_singlet_myelo_heatmap,
  features = features, assay = "RNA"
)

DoHeatmap(
  seu_singlet_myelo_heatmap,
  features = features,
  group.by = "ident", assay = "RNA"
)
```

### FeaturePlot of microglia/macrophages selected core markers
```{r, echo=FALSE, message=FALSE, fig.width=7, fig.height=6}
core_list <- list(
  mg_core = c("P2RY12", "TMEM119", "CX3CR1"),
  mac_core <- c("CD163", "TGFBI", "CD68", "AIF1")
)

map(
  core_list,
  ~ FeaturePlot(
    seu_singlet_myelo,
    features = .x,
    reduction = "wnn.umap",
    pt.size = 1,
    label = FALSE
  ) & mycolor2 & NoAxes()
)
```

### DensityPlot of microglia/macrophages selected core markers
```{r, echo=FALSE, message=FALSE}
map(
  core_list,
  ~ plot_density(
    seu_singlet_myelo,
    features = .x,
    size = 1.5,
    joint = "TRUE",
    reduction = "wnn.umap"
  ) & NoAxes()
)
```

# ADT features
```{r, eval=FALSE, echo=FALSE}
DefaultAssay(seu_singlet_myelo) <- "ADT.dsb.threshold"
features <- get_features(seu_singlet_myelo, "ADT.dsb.threshold")
length(features)
f1 <- features[1:10]
f2 <- features[11:21]
f_list <- list(f1, f2)

for (i in seq_along(f_list)) {
  seu_FeaturePlot(
    seu_singlet_myelo, "ADT.dsb.threshold",
    reduction = "wnn.umap",
    color = inferno,
    features = f_list[[i]],
    max.cutoff = NA,
    show = FALSE,
    save_path = paste0("singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_", i), # nolint
    ggheight = 20,
    ggwidth = 20
  ) & NoAxes()
}

for (i in seq_along(f_list)) {
  plot <- plot_density(
    seu_singlet_myelo,
    features = f_list[[i]],
    reduction = "wnn.umap",
    size = 1.5,
  ) & NoAxes() & inferno

  ggsave(
    paste0("singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_density_", i, ".png"),
    plot,
    height = 20,
    width = 20
  )
}
DefaultAssay(seu_singlet_myelo) <- "RNA"
```

## load adt features graphs
### feature plot
```{r, echo=FALSE, fig.cap="ADT filtered features",}
knitr::include_graphics("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_1.png") # nolint

knitr::include_graphics("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_2.png") # nolint
```

### density plot
```{r, echo=FALSE, fig.cap="ADT filtered features, density plot",}
knitr::include_graphics("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_density_1.png") # nolint

knitr::include_graphics("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-25_sct_dsb-threshold_myelo/panel_density_2.png") # nolint
```

# Pathway analysis
## GSVA (GSVA using gsva package)
The GSVA package provides an implementation of this approach for the following methods:
* gsva (HÃ¤nzelmann, Castelo, and Guinney 2013). This is the default method of the package and similarly to ssGSEA, is a non-parametric method that uses the empirical CDFs of gene expression ranks inside and outside the gene set, but it starts by calculating an expression-level statistic that brings gene expression profiles with different dynamic ranges to a common scale.
### GSVA msigdb Hallmark collection = 0 DE pathways
```{r, eval=FALSE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "H",
  gsva_type = "gsva"
)
# as it is 0 pathways here, code for calculating and  plotting is removed
```

### GSVA msgidb C2 collection
```{r, eval=FALSE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "C2",
  gsva_type = "gsva"
)

saveRDS(gsva_es, "/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/21-Feb-25_gsva_es_myelo_C2.rds") # nolint
```

```{r, echo=FALSE, warning=FALSE}
# as gsva for C2 takes time, save and load .rds with gsva calculated
gsva_es <- readRDS("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/21-Feb-25_gsva_es_myelo_C2.rds") # nolint

gsva_cx <- "GSVA_C2"
seu_singlet_myelo[[gsva_cx]] <- Seurat::CreateAssayObject(data = gsva_es)
Seurat::DefaultAssay(seu_singlet_myelo) <- gsva_cx
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
markers <- Seurat::FindAllMarkers(
  seu_singlet_myelo,
  assay = "GSVA_C2",
  only.pos = FALSE
)
markers_up <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = 20, wt = avg_log2FC)
markers_down <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = -20, wt = avg_log2FC)
top_markers <- bind_rows(markers_up, markers_down) |>
  mutate(upordown = ifelse(avg_log2FC > 0, "up", "down"))
```

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
datatable(top_markers)

seu_gsva_print_plot(
  top_markers,
  "GSVA top 40(20up+20down) DE (wilcox) C2 pathways in Cluster "
)
```

## zscore method (using gsva package)
### zscore msigdb Hallmark collection
```{r, eval=FALSE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "H",
  gsva_type = "zscore"
)
```

```{r, echo=FALSE, warning=FALSE}
# as gsva for C2 takes time, save and load .rds with gsva calculated
gsva_cx <- "GSVA_H_zscore"
seu_singlet_myelo[[gsva_cx]] <- Seurat::CreateAssayObject(data = gsva_es)
Seurat::DefaultAssay(seu_singlet_myelo) <- gsva_cx
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
markers <- Seurat::FindAllMarkers(
  seu_singlet_myelo,
  assay = gsva_cx,
  only.pos = FALSE
)
markers_up <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = 20, wt = avg_log2FC)
markers_down <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = -20, wt = avg_log2FC)
top_markers <- bind_rows(markers_up, markers_down) |>
  mutate(upordown = ifelse(avg_log2FC > 0, "up", "down"))
```

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
datatable(top_markers)

seu_gsva_print_plot(
  top_markers,
  "zscoreGSVA top 40(20up+20down) DE (wilcox) Hallmark pathways in Cluster "
)
```

### zscore msigdb C2 collection
```{r, eval=TRUE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "C2",
  gsva_type = "zscore"
)
```

```{r, echo=FALSE, warning=FALSE}
gsva_cx <- "GSVA_C2_zscore"
seu_singlet_myelo[[gsva_cx]] <- Seurat::CreateAssayObject(data = gsva_es)
Seurat::DefaultAssay(seu_singlet_myelo) <- gsva_cx
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
markers <- Seurat::FindAllMarkers(
  seu_singlet_myelo,
  assay = gsva_cx,
  only.pos = FALSE
)
markers_up <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = 20, wt = avg_log2FC)
markers_down <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = -20, wt = avg_log2FC)
top_markers <- bind_rows(markers_up, markers_down) |>
  mutate(upordown = ifelse(avg_log2FC > 0, "up", "down"))
```

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
datatable(top_markers)

seu_gsva_print_plot(
  top_markers,
  "zscoreGSVA top 40(20up+20down) DE (wilcox) Hallmark pathways in Cluster "
)
```

## PLAGE (using GSVA package)
*plage (Tomfohr, Lu, and Kepler 2005). Pathway level analysis of gene expression (PLAGE) standardizes expression profiles over the samples and then, for each gene set, it performs a singular value decomposition (SVD) over its genes. The coefficients of the first right-singular vector are returned as the estimates of pathway activity over the samples. Note that, because of how SVD is calculated, the sign of its singular vectors is arbitrary.

## PLAGE msigdb Hallmark 
```{r, eval=TRUE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "H",
  gsva_type = "plage"
)
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
# as gsva for C2 takes time, save and load .rds with gsva calculated
gsva_cx <- "GSVA_H_plage"
seu_singlet_myelo[[gsva_cx]] <- Seurat::CreateAssayObject(data = gsva_es)
Seurat::DefaultAssay(seu_singlet_myelo) <- gsva_cx
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
markers <- Seurat::FindAllMarkers(
  seu_singlet_myelo,
  assay = gsva_cx,
  only.pos = FALSE
)
markers_up <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = 20, wt = avg_log2FC)
markers_down <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = -20, wt = avg_log2FC)
top_markers <- bind_rows(markers_up, markers_down) |>
  mutate(upordown = ifelse(avg_log2FC > 0, "up", "down"))
```

```{r, eval=TRUE, fig.width=15, fig.height=8}
datatable(top_markers)

seu_gsva_print_plot(
  top_markers,
  "plageGSVA top 40(20up+20down) DE (wilcox) Hallmark pathways in Cluster "
)
```

## PLAGE msigdb C2
```{r, eval=FALSE, message=FALSE, echo=FALSE, warning=FALSE}
gsva_es <- seu_gsva(
  seu_singlet_myelo,
  category = "C2",
  gsva_type = "plage"
)

# this one also takes a lot of time. save and load the results.
saveRDS(gsva_es, "/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/21-Feb-25_gsva_es_myelo_C2_plage.rds") # nolint
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
# as gsva for C2 takes time, save and load .rds with gsva calculated
gsva_es <- readRDS("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/RDS/21-Feb-25_gsva_es_myelo_C2_plage.rds") # nolint

gsva_cx <- "GSVA_C2_plage"
seu_singlet_myelo[[gsva_cx]] <- Seurat::CreateAssayObject(data = gsva_es)
Seurat::DefaultAssay(seu_singlet_myelo) <- gsva_cx
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE}
markers <- Seurat::FindAllMarkers(
  seu_singlet_myelo,
  assay = gsva_cx,
  only.pos = FALSE
)
markers_up <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = 20, wt = avg_log2FC)
markers_down <- markers |>
  filter(p_val_adj < 0.05) |>
  group_by(cluster) |>
  arrange(desc(avg_log2FC)) |>
  top_n(n = -20, wt = avg_log2FC)
top_markers <- bind_rows(markers_up, markers_down) |>
  mutate(upordown = ifelse(avg_log2FC > 0, "up", "down"))
```

```{r, eval=TRUE, fig.width=15, fig.height=8}
datatable(top_markers)

seu_gsva_print_plot(
  top_markers,
  "plageGSVA top 40(20up+20down) DE (wilcox) C2 gene sets in Cluster "
)
```

# GSEA (using fgsea package)
```{r}
DefaultAssay(seu_singlet_myelo) <- "RNA"
gsea_markers <- FindAllMarkers(
  seu_singlet_myelo,
  only.pos = FALSE,
  logfc.threshold = 0,
  min.pct = 0
)
head(gsea_markers)
```

## Hallmark 
```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
for (i in unique(gsea_markers$cluster)) {
  cluster_markers <- gsea_markers |>
    filter(p_val_adj < 0.05) |>
    filter(cluster == i) |>
    mutate(myscore = sign(avg_log2FC) * abs(avg_log2FC) * -log10(p_val_adj))

  ranked_genes <- cluster_markers$myscore
  names(ranked_genes) <- cluster_markers$gene
  ranked_genes <- sort(ranked_genes, decreasing = TRUE)
  ranked_genes <- ranked_genes[is.finite(ranked_genes)]

  top_pathways <- seu_gsea(ranked_genes, "H")
  seu_gsea_plot(top_pathways, "H", i)
}
```

## C2
```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
for (i in unique(gsea_markers$cluster)) {
  cluster_markers <- gsea_markers |>
    filter(p_val_adj < 0.05) |>
    filter(cluster == i) |>
    mutate(myscore = sign(avg_log2FC) * abs(avg_log2FC) * -log10(p_val_adj))

  ranked_genes <- cluster_markers$myscore
  names(ranked_genes) <- cluster_markers$gene
  ranked_genes <- sort(ranked_genes, decreasing = TRUE)
  ranked_genes <- ranked_genes[is.finite(ranked_genes)]

  top_pathways <- seu_gsea(ranked_genes, "C2")
  seu_gsea_plot(top_pathways, "C2", i)
}
```

# Trajectory analysis with Monocle3
```{r, eval=FALSE, echo=FALSE}
## Convert SingleCellExperiment object to CellDataSet
cds <- SeuratWrappers::as.cell_data_set(seu_singlet_myelo)

## Extract UMAP embeddings from Seurat
umap_coords <- Embeddings(seu_singlet_myelo, reduction = "wnn.umap")

## Add UMAP embeddings to Monocle3
reducedDims(cds)$UMAP <- umap_coords

## Cluster the cells
cds <- cluster_cells(cds, reduction_method = "UMAP")

## Plot cells
cds <- learn_graph(cds, use_partition = TRUE)

plot_cells(cds,
  label_groups_by_cluster = FALSE,
  label_leaves = FALSE,
  label_branch_points = FALSE,
  color_cells_by = "seurat_clusters"
)

cds <- order_cells(cds)

plot <- plot_cells(cds,
  color_cells_by = "pseudotime",
  label_cell_groups = FALSE,
  label_leaves = FALSE,
  label_branch_points = FALSE,
  graph_label_size = 1.5,
)
ggsave("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-2025_seu_myelo_mococle3.png", plot)
```

```{r, echo=FALSE, fig.cap="Monocle3 result",}
knitr::include_graphics("/mnt/sda4/singleCell_LAB/HitGlio_sample1-test/plots/10-feb-2025_seu_myelo_mococle3.png") # nolint
```

# Check for tumor cells with InferCNV
```{r}

```

# Ligand-Receptor analysis /myeloid only/ (cellchat)
```{r, eval=FALSE}
data.input <- seu_singlet_myelo[["RNA"]]$data
labels <- Idents(seu_singlet_myelo)
meta <- data.frame(labels = labels, row.names = names(labels)) # create a dataframe of the cell labels

library(CellChat)
options(stringsAsFactors = FALSE)

cellchat <- createCellChat(
  object = seu_singlet_myelo,
  group.by = "ident", assay = "RNA"
)
CellChatDB <- CellChatDB.human
cellchat@DB <- CellChatDB

# subset the expression data of signaling genes for saving computation cost
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 24) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)

cellchat <- computeCommunProb(cellchat, type = "triMean")

cellchat <- computeCommunProbPathway(cellchat)

cellchat <- aggregateNet(cellchat)


groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1, 2), xpd = TRUE)
netVisual_circle(
  cellchat@net$count,
  vertex.weight = groupSize,
  weight.scale = T,
  label.edge = F,
  title.name = "Number of interactions"
)

netVisual_circle(
  cellchat@net$weight,
  vertex.weight = groupSize,
  weight.scale = T,
  label.edge = F,
  title.name = "Interaction weights/strength"
)

mat <- cellchat@net$weight
par(mfrow = c(3, 4), xpd = TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```





# testing ground
```{r, eval = FALSE, echo = FALSE}
top_x_inspect <- markers_genes %>%
  group_by(cluster) %>%
  slice_min(p_val_adj, n = 100, with_ties = FALSE)

top_x_inspect |>
  filter(cluster == 4) |>
  select(gene) |>
  as.vector()

FeaturePlot(
  seu_singlet_myelo,
  # feature = c("TMEM119", "P2RY12", "CD163", "CCR2"),
  # feature = c("TREM2", "CD84"),
  feature = c("CCR5"),
  reduction = "wnn.umap",
  pt.size = 1,
  alpha = 0.8,
) & mycolor3 & NoAxes()

mycolor3 <- viridis::scale_color_viridis(
  option = "viridis",
  alpha = 1,
  begin = 0,
  end = 1,
  direction = 1
)

plot_density(
  seu_singlet_myelo,
  c("GFAP", "PTPRZ1", "SCG3", "SOX2", "SOX6", "BCAN", "NOVA1", "NRCAM", "TUBB2B"),
  slot = "data",
  size = 1.5,
  joint = "TRUE",
  reduction = "wnn.umap"
) & NoAxes()

Idents(seu_singlet_myelo) <- group_by
VlnPlot(
  seu_singlet_myelo,
  "GFAP"
  # "SOX2"
)

glioblastoma_proneural <- c(
  "ABAT", "ACTR1A", "ALCAM", "AMOTL2", "KDM1A", "ARHGEF9", "ASCL1", "ATAD5", "ATP1A3", "ADGRB3",
  "BCAN", "BCL7A", "BCOR", "BEX1", "C1QL1", "INAVA", "MIR9-1HG", "FERMT1", "ATAT1", "CA10",
  "CAMSAP2", "CASK", "CBX1", "CDC25A", "CDC7", "CDK5R1", "CDKN1B", "CHD7", "CKB", "CLASP2",
  "CLGN", "CNTN1", "CRB1", "CRMP1", "CSNK1E", "CSPG5", "CXXC4", "DBN1", "DCX", "DGKI", "DLL3",
  "DNM3", "DPF1", "DPP6", "DPYSL4", "DUSP26", "E2F3", "EPB41", "EPHB1", "ERBB3", "FAM110B",
  "MVB12B", "NKAIN1", "FBXO21", "FGF9", "FHOD3", "FLRT1", "FXYD6", "GABRA3", "GADD45G", "GNG4",
  "GPM6A", "GPR17", "LPAR4", "GRIA2", "GRID2", "GSK3B", "GSTA4", "HDAC2", "HMGB3", "JPT1",
  "HOXD3", "PLAAT1", "CILK1", "IL1RAPL1", "ZC4H2", "KIF21B", "KLRC3", "KLRC4", "KLRK1", "ZNF821",
  "REXO5", "ADGRL3", "LRP6", "LRRTM4", "MAP2", "MAPT", "MARCKS", "MARCKSL1", "MAST1", "MATR3",
  "MCM10", "MLLT11", "MMP15", "MMP16", "MPPED2", "MTSS1", "MYB", "MYO10", "KAT7", "MYT1",
  "NCALD", "NCAM1", "NKX2-2", "NLGN3", "NOL4", "NR0B1", "NRXN1", "NRXN2", "OLIG2", "P2RX7",
  "PAFAH1B3", "PAK3", "PAK5", "PCDH11X", "PCDH11Y", "PDE10A", "PELI1", "PFN2", "JADE3", "PHLPP1",
  "PLCB4", "PODXL2", "PPM1D", "PPM1E", "PURG", "RAB33A", "RAD21", "RALGPS1", "RALGPS2", "RAP2A",
  "RBPJ", "REEP1", "RUFY3", "SATB1", "SCG3", "SCN3A", "SEC61A2", "SEZ6L", "SLC1A1", "SLCO5A1",
  "ARHGAP33", "SORCS3", "SOX10", "SOX11", "SOX2", "SOX4", "SPTBN2", "SRGAP3", "STMN1", "STMN4",
  "TAF5", "TMCC1", "TMEFF1", "RNFT2", "TMEM35A", "TMSB15A", "CELF3", "TOP2B", "TOPBP1", "TOX3",
  "TSPAN3", "TTC3", "TTYH1", "VAX2", "VEZF1", "WASF1", "DCAF7", "YPEL1", "ZEB2", "ZNF184",
  "ZNF248", "ZNF286A", "ZNF510", "ZFP69B", "ZNF711", "ZNF804A"
)
```


```{r}
top_x <- markers_genes %>%
  group_by(cluster) %>%
  slice_min(p_val_adj, n = 50, with_ties = FALSE)

cl3 <- top_x |>
  filter(cluster == myelo_cluster_annotation[3])

print(cl3, n = 100)
```


# transfer labels to the original seurat object
```{r}

```